### DISCUSSION TITLE
Why is EGID marked as DoNotSerialize

### DISCUSSION THREAD

**Discussion by simeonradivoev** [sebas77/Svelto.ECS#99] [Q&A]
Why is EGID marked as DoNotSerialize yet has Serializable

**Comment by sebas77** [comment]
EGID is not deterministic between sessions. I.E.: if you store it in a save file and load it after code changed it won't make sense

**Reply by simeonradivoev** [reply]
in what case are you not restoring entity IDs? It seems like the most common use case to store a reference to an entity.

**Reply by sebas77** [reply]
you shouldn't ever need to use EGID directly, you should use EntityReference instead. This is just because EntityReference is group agnostic (the reference is valid even if the group changes). However, serialization is a different problem, as I explained Svelto cannot guarantee that the deserialized EGID has any meaning (the entity is in the database and is the one you expect with that EGID).

Serialization is a complex topic, you can study the GroupHashMap and the SerializableEntityDescriptor to see how some problems have been solved, however, runtime IDs would make sense only if you can guarantee you can create the entities back with those IDs again (which is not impossible if you design the game having this as a rule)

**Reply by simeonradivoev** [reply]
So the solution is to never use EGID or EntityReference in components. You have to use a custom type/id. as I can see EntityReference is also marked as DoNotSerialize

**Reply by sebas77** [reply]
yeah that's what we did, although I don't remember exactly the mechanism. For network we used to have a NetID that was able to bind a deterministic ID to EGID on client and server, while for the serialization on disk I am not sure how we solved it maybe @Grigler can shed some light.
